File: packages/core/src/recommender.ts
/**
 * recommender.ts — Program recommendation engine
 *
 * Matches a student's RIASEC profile to SNIES programs using
 * the RIASEC → CINE field mapping. Supports enrollment-weighted
 * priors that surface hidden-gem programs.
 */

import { cosineSimilarity } from "./scoring.js";
import { loadMapping } from "./programs.js";
import { RIASEC_TYPES } from "./types.js";
import type {
  RIASECProfile,
  RIASECType,
  Program,
  TypeMapping,
  RecommendConfig,
  ProgramFilters,
  Recommendation,
} from "./types.js";

const VIRTUAL_MODALITIES = new Set([
  "Virtual",
  "A distancia",
  "Virtual-A distancia",
]);

/**
 * Build a RIASEC profile vector for a given CINE broad field,
 * based on the mapping weights.
 */
function fieldToProfile(
  cineAmplio: string,
  mapping: Record<RIASECType, TypeMapping>
): RIASECProfile {
  const profile: RIASECProfile = { R: 0, I: 0, A: 0, S: 0, E: 0, C: 0 };

  for (const type of RIASEC_TYPES) {
    const entry = mapping[type];
    for (const field of entry.fields) {
      if (field.cine_amplio === cineAmplio) {
        profile[type] = field.weight;
      }
    }
  }

  return profile;
}

/** Apply filters to a program list */
function applyFilters(programs: Program[], filters?: ProgramFilters): Program[] {
  if (!filters) return programs;

  let result = programs;

  if (filters.active !== undefined) {
    const status = filters.active ? "Activo" : "Inactivo";
    result = result.filter((p) => p.estado === status);
  }

  if (filters.departments?.length) {
    const depts = new Set(filters.departments);
    result = result.filter((p) => depts.has(p.departamento));
  }

  if (filters.nivel_formacion?.length) {
    const niveles = new Set(filters.nivel_formacion);
    result = result.filter((p) => niveles.has(p.nivel_formacion));
  }

  if (filters.modalidad?.length) {
    const mods = new Set(filters.modalidad);
    result = result.filter((p) => mods.has(p.modalidad));
  }

  if (filters.sector?.length) {
    const sectors = new Set(filters.sector);
    result = result.filter((p) => sectors.has(p.sector));
  }

  return result;
}

/**
 * Compute program counts per CINE broad field (for enrollment-based priors).
 * Uses program count as a proxy for enrollment when actual enrollment data
 * is not available.
 */
function computeFieldCounts(programs: Program[]): Map<string, number> {
  const counts = new Map<string, number>();
  for (const p of programs) {
    const field = p.cine_amplio;
    if (field) {
      counts.set(field, (counts.get(field) ?? 0) + 1);
    }
  }
  return counts;
}

/**
 * Recommend programs based on RIASEC profile.
 *
 * score(program) = similarity(student_profile, program_cine_vector)
 *               × prior(program)
 *
 * prior(program) = base
 *               × (1 + enrollmentWeight × log(national / field_count))
 *               × (1 + regionalBoost × isInRegion)
 *               × (1 + virtualBoost × isVirtual)
 */
export function recommend(config: RecommendConfig): Recommendation[] {
  const {
    profile,
    programs,
    filters,
    limit = 20,
  } = config;

  const mapping = config.mapping ?? loadMapping();
  const priors = {
    enrollmentWeight: config.priors?.enrollmentWeight ?? -0.3,
    regionalBoost: config.priors?.regionalBoost ?? 1.0,
    virtualBoost: config.priors?.virtualBoost ?? 1.0,
  };

  const filtered = applyFilters(programs, filters);
  const fieldCounts = computeFieldCounts(programs); // use full dataset for counts
  const totalPrograms = programs.filter((p) => p.estado === "Activo").length;

  // Pre-compute field profiles
  const fieldProfileCache = new Map<string, RIASECProfile>();
  for (const field of fieldCounts.keys()) {
    fieldProfileCache.set(field, fieldToProfile(field, mapping));
  }

  const regionalDepts = filters?.departments
    ? new Set(filters.departments)
    : null;

  const results: Recommendation[] = [];

  for (const program of filtered) {
    const fieldProfile = fieldProfileCache.get(program.cine_amplio);
    if (!fieldProfile) continue;

    // Profile similarity (cosine)
    const similarity = cosineSimilarity(profile, fieldProfile);
    if (similarity === 0) continue;

    // Enrollment-based prior
    const fieldCount = fieldCounts.get(program.cine_amplio) ?? 1;
    const enrollmentFactor =
      1 + priors.enrollmentWeight * Math.log(totalPrograms / fieldCount);

    // Regional boost
    const isRegional = regionalDepts?.has(program.departamento) ? 1 : 0;
    const regionalFactor = 1 + priors.regionalBoost * isRegional;

    // Virtual boost
    const isVirtual = VIRTUAL_MODALITIES.has(program.modalidad) ? 1 : 0;
    const virtualFactor = 1 + priors.virtualBoost * isVirtual;

    const priorAdjustment = enrollmentFactor * regionalFactor * virtualFactor;
    const score = similarity * priorAdjustment;

    // Find which RIASEC types match this field
    const matchingTypes = RIASEC_TYPES.filter(
      (t) => fieldProfile[t] > 0
    );

    results.push({
      program,
      score,
      matchDetails: {
        profileSimilarity: similarity,
        priorAdjustment,
        matchingTypes,
      },
    });
  }

  // Sort by score descending, deduplicate by program name + institution
  results.sort((a, b) => b.score - a.score);

  // Deduplicate: keep highest-scoring entry per unique program
  const seen = new Set<string>();
  const deduped: Recommendation[] = [];
  for (const r of results) {
    const key = `${r.program.codigo_snies}`;
    if (!seen.has(key)) {
      seen.add(key);
      deduped.push(r);
    }
  }

  return deduped.slice(0, limit);
}

File: python/src/riasec_co/recommender.py
"""Program recommendation engine."""

from __future__ import annotations

import math

import polars as pl

from riasec_co.data import Programs, load_mapping
from riasec_co.scoring import RIASEC_TYPES, cosine_similarity

VIRTUAL_MODALITIES = {"Virtual", "A distancia", "Virtual-A distancia"}


def _field_to_profile(
    cine_amplio: str,
    mapping: dict[str, dict],
) -> dict[str, float]:
    """Build a RIASEC profile vector for a CINE broad field."""
    profile = {t: 0.0 for t in RIASEC_TYPES}
    for riasec_type in RIASEC_TYPES:
        entry = mapping.get(riasec_type, {})
        for field_entry in entry.get("fields", []):
            if field_entry["cine_amplio"] == cine_amplio:
                profile[riasec_type] = field_entry["weight"]
    return profile


def recommend(
    profile: dict[str, float],
    programs: pl.DataFrame | None = None,
    mapping: dict[str, dict] | None = None,
    *,
    departments: list[str] | None = None,
    active_only: bool = True,
    nivel_formacion: list[str] | None = None,
    modalidad: list[str] | None = None,
    enrollment_weight: float = -0.3,
    regional_boost: float = 1.0,
    virtual_boost: float = 1.0,
    limit: int = 20,
) -> pl.DataFrame:
    """Recommend programs based on a RIASEC profile.

    Returns a Polars DataFrame with columns from the program data plus
    'score', 'similarity', and 'matching_types'.
    """
    if programs is None:
        programs = Programs.load()
    if mapping is None:
        mapping = load_mapping()

    df = programs

    # Apply filters
    if active_only:
        df = df.filter(pl.col("estado") == "Activo")
    if departments:
        df = df.filter(pl.col("departamento").is_in(departments))
    if nivel_formacion:
        df = df.filter(pl.col("nivel_formacion").is_in(nivel_formacion))
    if modalidad:
        df = df.filter(pl.col("modalidad").is_in(modalidad))

    # Pre-compute field profiles
    field_profiles: dict[str, dict[str, float]] = {}
    for field_name in df.get_column("cine_amplio").drop_nulls().unique().to_list():
        field_profiles[field_name] = _field_to_profile(field_name, mapping)

    # Compute field counts for enrollment-based priors
    all_active = programs.filter(pl.col("estado") == "Activo")
    total_programs = len(all_active)
    field_counts: dict[str, int] = {}
    for row in all_active.group_by("cine_amplio").len().iter_rows():
        field_counts[row[0]] = row[1]

    regional_set = set(departments) if departments else set()

    # Score each row
    scores = []
    similarities = []
    matching_types_list = []

    for row in df.iter_rows(named=True):
        cine = row.get("cine_amplio", "")
        field_profile = field_profiles.get(cine)

        if not field_profile or all(v == 0 for v in field_profile.values()):
            scores.append(0.0)
            similarities.append(0.0)
            matching_types_list.append("")
            continue

        sim = cosine_similarity(profile, field_profile)
        if sim == 0:
            scores.append(0.0)
            similarities.append(0.0)
            matching_types_list.append("")
            continue

        # Enrollment-based prior
        fc = field_counts.get(cine, 1)
        enrollment_factor = 1 + enrollment_weight * math.log(total_programs / fc)

        # Regional boost
        is_regional = 1 if row.get("departamento", "") in regional_set else 0
        regional_factor = 1 + regional_boost * is_regional

        # Virtual boost
        is_virtual = 1 if row.get("modalidad", "") in VIRTUAL_MODALITIES else 0
        virtual_factor = 1 + virtual_boost * is_virtual

        score = sim * enrollment_factor * regional_factor * virtual_factor
        matching = [t for t in RIASEC_TYPES if field_profile[t] > 0]

        scores.append(score)
        similarities.append(sim)
        matching_types_list.append(",".join(matching))

    result = df.with_columns(
        pl.Series("score", scores),
        pl.Series("similarity", similarities),
        pl.Series("matching_types", matching_types_list),
    )

    # Sort, deduplicate, re-sort, limit
    result = (
        result
        .sort("score", descending=True)
        .unique(subset=["codigo_snies"], keep="first")
        .sort("score", descending=True)
        .head(limit)
    )

    return result
